### PR TITLE
Fix iOS "Arithmetic Overflow"

### LINKED ISSUE #1872: App freeze on iOS with "arithmetic overflow" error in sentry swift
### Platform

Dart

### Obfuscation

Disabled

### Debug Info

Disabled

### Doctor

[✓] Flutter (Channel stable, 3.16.2, on macOS 14.2.1 23C71 darwin-arm64, locale
    en-IL)
    • Flutter version 3.16.2 on channel stable at
      /Users/zvikarp/.asdf/installs/flutter/3.16.2-stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 9e1c857886 (10 weeks ago), 2023-11-30 11:51:18 -0600
    • Engine revision cf7a9d0800
    • Dart version 3.2.2
    • DevTools version 2.28.3

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/zvikarp/Library/Android/sdk
    • Platform android-34, build-tools 32.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15C500b
    • CocoaPods version 1.15.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      17.0.7+0-17.0.7b1000.6-10550314)

[✓] VS Code (version 1.83.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.82.0

[✓] VS Code (version 1.86.0-insider)
    • VS Code at /Applications/Visual Studio Code - Insiders.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (3 available)
  *******

[✓] Network resources
    • All expected network resources are available.

• No issues found!

### Version

7.14.0

### Steps to Reproduce

I'm not sure, this issue started happening recently, it is hard for me to understand why it is happening, sometimes it works and sometimes I get an error in xcode:

recording_2024-02-07_9 -29-18

To get to this error I run the app and at some stage it just freezes, hot reload stops working and the error above appears in xcode. seems like that the result of the selected line is negative but we are using unit here and this is the cause of the error.

### Expected Result

Not to freeze

### Actual Result

App freezes

### Are you willing to submit a PR?

None

### PR DESCRIPTION
**Pull request by denrase**
## :scroll: Description

- Guard against "Arithmetic Overflow"
- Fix duplicate finish transaction in `SentryNavigatorObserver`

## :bulb: Motivation and Context

Closes #1872

## :pencil: Checklist

- [x] I reviewed submitted code
- [ ] I added tests to verify changes
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPii` is enabled
- [ ] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: flutter/lib/src/navigation/sentry_navigator_observer.dart [hunk 1]
```diff
@@ -241,8 +242,13 @@ class SentryNavigatorObserver extends RouteObserver<PageRoute<dynamic>> {
   }
 
   Future<void> _finishTransaction() async {
+    final transaction = _transaction;
+    if (transaction == null || transaction.finished) {
+      return;
```
Comment (buenaflor) [inline flutter/lib/src/navigation/sentry_navigator_observer.dart:248]: what happens if we finish an already finished transaction again?

**Review by buenaflor — APPROVED** [review]
👍

**Comment by buenaflor** [comment]
Any tests we can add that make sense? Ideally I'd like to have changes be tested

File: flutter/lib/src/navigation/sentry_navigator_observer.dart (hunk 1 above)
Comment (denrase) [reply flutter/lib/src/navigation/sentry_navigator_observer.dart:248]: idk what happens on the sdk side, i would assume we handle this correctly, but the callback from above was called again, resulting in an additional bridge call

**Comment by denrase** [comment]
@buenaflor I will check regarding tests, also have failing ones.

**Comment by denrase** [comment]
@buenaflor Fixed the test and also added a test that multiple calls to `didPop` do not make multiple finish calls on the transaction. Please feel free to merge when CI passes. 🙇‍♂️

**Pull request merged**